### PR TITLE
Queue system for datastore

### PR DESCRIPTION
**Pull request by ketwaroo**
PQDC task

This should theoretically work and work without changing the external datastore packages.

Requires a cache rebuild for drupal 8 to discover new queue worker plugins.

At time of writing, it seems only the drush command `dkan-datastore:import` will actually call the import. There doesn't seem anywhere it could be triggered by the user via the UI

# Testing

Create a new dataset by POST to `/api/v1/dataset` It requires an API User to be authenticated by basic authentication.

The post body that should work:

```json
{
    "identifier": "23ew-n7w9",
    "accessLevel": "public",
    "landingPage": "https://data.medicare.gov/d/23ew-n7w9",
    "bureauCode": [
        "009:38"
    ],
    "issued": "2018-10-17",
    "modified": "2019-01-26T20:16:07",
    "keyword": [
        "dialysis facilities",
        "quality measures"
    ],
    "contactPoint": {
        "fn": "Dialysis Facility Compare",
        "hasEmail": "mailto:DialysisData@umich.edu"
    },
    "publisher": {
        "name": "Centers for Medicare & Medicaid Services (CMS)",
        "identifier": "centersformedicaremedicaidservicescms",
        "created": "2019-01-26T20:16:07",
        "modified": "2019-01-26T20:16:07"
    },
    "description": "A list of all dialysis facilities registered with Medicare that includes addresses and phone numbers, as well as services and quality of care provided.",
    "title": "Dialysis Facility Compare - Listing by Facility",
    "programCode": [
        "009:000"
    ],
    "distribution": [
        {
            "downloadURL": "https://data.medicare.gov/api/views/23ew-n7w9/rows.csv?accessType=DOWNLOAD",
            "mediaType": "text/csv",
            "title": "csv",
            "format": "csv",
            "identifier": "https-data-medicare-gov-api-views-23ew-n7w9-csv"
        },
        {
            "downloadURL": "https://data.medicare.gov/api/views/23ew-n7w9/rows.rdf?accessType=DOWNLOAD",
            "mediaType": "application/rdf+xml",
            "title": "rdf+xml",
            "format": "rdf+xml",
            "identifier": "https-data-medicare-gov-api-views-23ew-n7w9-rdf-xml"
        },
        {
            "downloadURL": "https://data.medicare.gov/api/views/23ew-n7w9/rows.json?accessType=DOWNLOAD",
            "mediaType": "application/json",
            "title": "json",
            "format": "json",
            "identifier": "https-data-medicare-gov-api-views-23ew-n7w9-json"
        },
        {
            "downloadURL": "https://data.medicare.gov/api/views/23ew-n7w9/rows.xml?accessType=DOWNLOAD",
            "mediaType": "application/xml",
            "title": "xml",
            "format": "xml",
            "identifier": "https-data-medicare-gov-api-views-23ew-n7w9-xml"
        }
    ],
    "theme": [
        "Dialysis Facility Compare"
    ],
    "created": "2019-01-26T20:16:07"
}
``` 

This *should* create a dataset with identifier `23ew-n7w9`

use the drush command to import the resource using a deferred queue.
```
drush dkan-datastore:import "23ew-n7w9" true
```

It will help to have something like https://www.drupal.org/project/queue_ui installed to easily inspect the state of the queue.

Queue items will get automatically processed at each cron run. Obviously cron must either be set up to run periodically via system scheduler/crontab or launch manually.

### CODE REVIEW

File: modules/custom/dkan_datastore/src/Plugin/QueueWorker/DatastoreImportQueue.php [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+   * @param string $filePath
+   * @param array $importConfig
+   * @return \Dkan\Datastore\Manager\IManager
+   */
+  protected function getManager($uuid, $resourceId, string $filePath, array $importConfig) {
+    /** @var \Drupal\dkan_datastore\Manager\DatastoreManagerBuilder $managerBuilder */
+    $managerBuilder = \Drupal::service('dkan_datastore.manager.datastore_manager_builder');
+
+    /** @var \Dkan\Datastore\Manager\IManager $manager */
+    $manager = $managerBuilder->setResource($resourceId, $filePath)
```
Comment (fmizzell) [inline modules/custom/dkan_datastore/src/Plugin/QueueWorker/DatastoreImportQueue.php]: @ketwaroo I believe that once you set the resource, the UUID is ignored by the build process, but we still have to pass that information just to ignore it. Maybe we should have two methods to set the resource, the one that exists and a second one that takes a UUID and sets it in that way. The resource is required, so maybe we just throw an exception if someone calls build without setting a resource previously.

File: modules/custom/dkan_datastore/src/Util.php [hunk 2]
```diff
@@ -3,37 +3,42 @@
...
-    $bin_storage = new LockableBinStorage("dkan_datastore", new Locker("dkan_datastore"), \Drupal::service('dkan_datastore.variable'));
-    $factory = new Factory($resource, $provider, $bin_storage, $database);
-
-    return $factory->get();
+  public static function getDatastoreManager(string $uuid) : IManager {
+
+    /** @var Manager\DatastoreManagerBuilder $builder */
+    $builder = \Drupal::service('dkan_datastore.manager.datastore_manager_builder');
+    return $builder->build($uuid);
+
```
Comment (fmizzell) [inline modules/custom/dkan_datastore/src/Util.php]: @ketwaroo We can go ahead and remove these comments.

File: modules/custom/dkan_datastore/src/Plugin/QueueWorker/DatastoreImportQueue.php (hunk 1 above)
Comment (ketwaroo) [reply modules/custom/dkan_datastore/src/Plugin/QueueWorker/DatastoreImportQueue.php]: Ah. yes. I see what you mean. I'll amend shortly

In queue mode, the uuid is passed mainly so we have something to name the remporary file and to record it in the logs for easier lookup.

File: modules/custom/dkan_datastore/src/Util.php (hunk 2 above)
Comment (ketwaroo) [reply modules/custom/dkan_datastore/src/Util.php]: fixed in 41751bae2e15caae5a840a6a3ad9834172e61879

**Pull request merged**